module/x/gravity/keeper/msg_server.go

```diff
@@ -29,45 +29,85 @@ var _ types.MsgServer = msgServer{}
 func (k msgServer) SetDelegateKeys(c context.Context, msg *types.MsgDelegateKeys) (*types.MsgDelegateKeysResponse, error) {
 	ctx := sdk.UnwrapSDKContext(c)
 
-	// these errors are checked in validate basic
-	val, _ := sdk.ValAddressFromBech32(msg.ValidatorAddress)
-	orch, _ := sdk.AccAddressFromBech32(msg.OrchestratorAddress)
-	eth := common.HexToAddress(msg.EthereumAddress)
+	valAddr, err := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	orchAddr, err := sdk.AccAddressFromBech32(msg.OrchestratorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	ethAddr := common.HexToAddress(msg.EthereumAddress)
 
 	// ensure that the validator exists
-	if k.Keeper.StakingKeeper.Validator(ctx, val) == nil {
-		return nil, sdkerrors.Wrap(stakingtypes.ErrNoValidatorFound, val.String())
+	if k.Keeper.StakingKeeper.Validator(ctx, valAddr) == nil {
+		return nil, sdkerrors.Wrap(stakingtypes.ErrNoValidatorFound, valAddr.String())
+	}
+
+	// NOTE: If a validator previously set their delegate keys, then there must
+	// exist Ethereum and orchestrator keys for that validator. In this case, we
+	// delete the values and allow the validator to update them, being sure to
+	// reset the keys to their original values in the case of either of the new
+	// keys existing.
+	//
+	// Otherwise, we treat the validator as it is setting their delegate keys for
+	// the first time and only erroring if either of the keys exist.
+
+	currEthAddr := k.GetValidatorEthereumAddress(ctx, valAddr)
+	if len(currEthAddr) > 0 {
```

Comment:
this will always return true. you should check if `currentEthAddr != common.Address{}`

Reply:
Ahh yes, I recall seeing this in the tests when I logged. Granted, it's harmless, but we could avoid a Delete call if we know it's empty. Is there an API we can use to determine if it's empty/zero?

Reply:
no, we just created a util function for this. You can also use `bytes.Equal`

Reply:
What's the util function?